### ISSUE TITLE
dart analyze mistake (Unnecessary cast) on int to num cast with var declaration

### ISSUE THREAD

**Issue by siahlooei** [dart-lang/sdk#54614]
This issues is related to:
* Dart analyzer and linter

Dart 3.2.3 (stable) (Tue Dec 5 17:58:33 2023 +0000) on "windows_x64"
locale is en-US

When i analyzed this simple code:
```dart
void main() {
  var varNum = 0 as num; // cast
  varNum = 2.3;
  print(varNum);
}
```
i got
```
dart analyze
Analyzing simple...

   info • main.dart:3:16 • Unnecessary cast. Try removing the cast. • unnecessary_cast

1 issue found.
```

but this is a necessary cast, because if i delete this cast the code is not compilable anymore.

**Comment by eernstg** [comment]
You might just as well use an explicit type annotation in the variable declaration:
```dart
void main() {
  num varNum = 0;
  varNum = 2.3;
  print(varNum);
}
```
The cast had no other purpose than changing the inferred type of the variable, and that's just a more complex way to make it a `num` variable when you could also write it directly.

**Comment by dgreensp** [comment]
This doesn't reproduce for me in Dart 3.3.2.

Plus, even if it did, it's arguably not a bug... it's the linter saying there is a better way to write something, and indeed there is, so...?

**Comment by srawlins** [comment]
Correct, we changed `unnecessary_cast` to only report when the cast type is the exact same as the expression's static type. Thank for catching, @dgreensp !
